### PR TITLE
fv3fit.train uses tfdataset instead of xr.Dataset sequences

### PR DESCRIPTION
**Pull request by mcgibbon**
This PR replaces the use of batches inside the fv3fit.train training code with tfdatasets. This should improve the performance of the data pipelining, reduce the amount of user code we have to rely on for data pipelining, and improve compatibility of fv3fit.train routines with emulation training code (which already uses tfdatasets).

Refactored public API:
- training in `fv3fit.train` is now in-memory by default, can be disabled with `in_memory: false` in the TrainingLoopConfig. A message prints when training with in_memory=True suggesting to disable it if training runs out of memory.
- "batch" for per-batch metrics now uses keras batches instead of loader batches
- Added clipping to convolutional training
- Default batch size for convolutional training is now 1
- Added clipping configuration to precipitative training
- refactored loss in precipitative training to be configured identically to the dense training
- TrainingLoopConfig/keras model training will now shuffle across a configurable number of samples (2,000,000 by default, approximately 144 timesteps worth of C48 columnwise data), instead of shuffling within single batches
- The serialized format of random forest predictors has changed, such that old models cannot be loaded by the updated code and vice-versa

Significant internal changes:
- Added train_column_model and ModelBuilder, which refactor the dense and precipitative models to re-use as much code as possible
- Added functions which operate on tfdatasets: pack_tfdataset, unpack_tfdataset
- refactored all training functions used by `fv3fit.train` to take `tf.data.Dataset` instead of `Sequence[xr.Dataset]` for their train_batches and validation_batches arguments.
- Added functions which operate on mappings of string to tensor: clip_sample, select_keys, apply_to_mapping,
- Added functions which operate on tensors: unpack_sample, ensure_nd
- Added function tfdataset_from_batches which converts batches to tfdataset
- The format used by EpochResult has changed from history: tf.keras.callbacks.History to a batch_logs and epoch_logs value centered around dictionaries of keras metrics, which better reflects the values given by keras to keras callbacks

- [x] Tests added

Resolves #<github issues> [JIRA-TAG]

(Delete this for the commit message)
You are encouraged to check the PR against the [Code Review Checklist](https://paper.dropbox.com/doc/Code-Review-Checklist--A4lKrs~xg7w5Gsb39N6JLNQoAg-IlsYffZgTwyKEylty7NhY).

### CODE REVIEW

File: external/fv3fit/fv3fit/keras/_models/precipitative.py
Comment (mcgibbon) [inline external/fv3fit/fv3fit/keras/_models/precipitative.py:187]:
```suggestion
        clip_config: configuration of input and output clipping of last dimension.
```

File: external/fv3fit/tests/test_tfdataset_ops.py
Comment (frodre) [inline external/fv3fit/tests/test_tfdataset_ops.py:82]: if switching to `_seq_to_tf_dataset` I don't think these transform-specific tests are really necessary.  Mostly testing that they work with `dataset.map` which is controlled by the user and how they order these operations.  If all the data changes happen w/in a composed function we don't have to worry about intermediate tf_dataset states.

**Review by frodre — CHANGES_REQUESTED** [review]
Thanks for the extensive refactor, @mcgibbon!  This goes a long way towards harmonizing what we're doing with the emulation, and hopefully helps with the optimization of training.

One main change I would consider is the usage of `_seq_to_tf_dataset` (would have to become public) in place of `batches_to_tf_dataset`, which would have implications for a few things.  

> This is very similar to _seq_to_tf_dataset in fv3fit.emulation.data.load. The main difference is instead of calling dataset .map for each individual transform outside, there is single composed transform (which is created via config) passed into the dataset generator.

This is a chance to move that emulation work into a shared location and use it for both cases.  Using composed functions sent into the transform generator works basically the same way but without intermediate tf_datasets, so it should be a relatively small change to how you've implemented it.  I think this may ake the tf_dataset-specific pack/unpack unnecessary, and we could mesh other transforms the emulation uses.  I think if this is the way forward, it should be a separate PR that moves the tf dataset creation out of emulation and adds any necessary transform changes / tests, and that we merge that PR before the current.
Very open to your thoughts on whether you think this would work/be good.

Other than that, I just wanted to get confirmation on the potential `EpochHistory` batch logs bug, make sure the convolutional `.numpy()`accesses are okay, and finally,  run actual model trainings using this before it gets merged (if not done already).  It's a big change so worth doing some double checks!

File: external/fv3fit/tests/test_tfdataset_ops.py
Comment (mcgibbon) [reply external/fv3fit/tests/test_tfdataset_ops.py:82]: This is testing things work testing, I think? For example, it's testing that the returned PackingInfo is accurate (most of the code below), and it's testing that the function applied inside `map` is accurate.

With some kind of Pipeline class, you still generally want unit tests to make sure each of the functions composing the Pipeline are correct.

```
def pack_tfdataset(
    data: tf.data.Dataset, variable_names: Sequence[str],
) -> Tuple[tf.data.Dataset, PackingInfo]:
    sample = next(iter(data))
    features = []
    for name in variable_names:
        if len(sample[name].shape) > 0:
            features.append(sample[name].shape[-1])
        else:
            features.append(1)
    packing_info = PackingInfo(names=variable_names, features=features)
    return (
        data.map(
            lambda x: tf.concat(list(x[name] for name in variable_names), axis=-1)
        ),
        packing_info,
    )
```

File: external/fv3fit/tests/training/test_train.py
Comment (mcgibbon) [reply external/fv3fit/tests/training/test_train.py:364]: Will do.

File: external/fv3fit/fv3fit/keras/_models/shared/callbacks.py
Comment (frodre) [reply external/fv3fit/fv3fit/keras/_models/shared/callbacks.py:34]: Maybe use `.get` and append `[]` like below.  If it's not really something that happens an empty list is fine too.

File: external/fv3fit/fv3fit/keras/_models/shared/training_loop.py
Comment (frodre) [reply external/fv3fit/fv3fit/keras/_models/shared/training_loop.py:73]: >The keras docs weren't clear about under what circumstances it gives None

Classic tf docs.

File: external/fv3fit/tests/test_tfdataset_ops.py
Comment (frodre) [reply external/fv3fit/tests/test_tfdataset_ops.py:82]: This just seemed like a lot of extra code work to make sure that the tf_dataset plays with the individual transforms. If using another composition method, it would take the tf_dataset-specific code out of the testing mix, which is nice, but I also see why the flexibility of using `map` for individual transforms is also nice.

File: external/fv3fit/fv3fit/tfdataset.py
Comment (frodre) [reply external/fv3fit/fv3fit/tfdataset.py:63]: > Does Dataset.from_tensor_slices work on Mapping[str, tf.Tensor]?

I believe it does.  Our entire workflow uses `Mapping[str, tf.Tensor]` as the default data.  Previously we would select into tuples of tensors, but Noah worked it out so that wasn't necessary.

> We can also talk about how best to apply transforms.

I agree that the current transform infrastructure is tailored for emulation-specific tasks.  The main things that are difficult to rectify is how to apply transforms in a general and flexible way that translates to the problem areas you highlighted.  I don't think emulation has these answered completely, but the `TensorTransform` ideas Noah recently contributed are promising for forward/backward ops (like packing, but maybe this too more of an edge case).

**Review by frodre — COMMENTED** [review]
I think after getting more context and a better understanding from your replies, the differences between using our dataset generator is not actually the target.  It's figuring out how to do transforms in a harmonious way, which it probably isn't the right time for.  I personally think transform compositions into a `.map`  is a little cleaner than sequential `.map` ops, but you rightly point out that it doesn't work well when more flexibility is needed.  Perhaps we just need a better definition for the types of ops we want pre/post-netcdfs, pre/post-training, and for prediction in our corrective/emulation ML before we can bring components together.

**Comment by mcgibbon** [comment]
I have to update the regtest outputs for sklearn training because the numpy version was incremented. Not sure when it was incremented or why they aren't already updated, if I pulled the increment in from master.

**Comment by nbren12** [comment]
> I have to update the regtest outputs for sklearn training because the numpy version was incremented. Not sure when it was incremented or why they aren't already updated, if I pulled the increment in from master.

This was expected...and required by bumping to a numpy that is compatible with tensorflow 2.8. FWIW, I think `git log constraints.txt` can help you understand why a certain file changed:

```
commit 97d2e459d917dc7d879c5405bd830b82ac8b1b82
Author: Noah D. Brenowitz <nbren12@gmail.com>
Date:   Mon Mar 7 21:42:09 2022 -0800

    fix dataflow submission
    
    The dependencies in `workflows/dataflow/setup.py` were incompatible with
    constraints.txt. Dataflow now supports running via a docker image, which will be
    easier to ensure a consistent environment.
    
    * switch to using docker for dataflow jobs
    * ensure dependencies are compatible with the versions installed in the dataflow
      base image
    
    These dependencies mostly match our stack, but did force some updates to
    libraries in the google stack like tensorflow and the google-cloud-* packages.
    
    Resolves #1678
```

**Review by frodre — APPROVED** [review]
LGTM, just a couple minor comments other than the failing np version test.  Thanks for addressing all the comments! I tried out this branch with the emulation training and everything worked out swell.

**Comment by mcgibbon** [comment]
> FWIW, I think git log constraints.txt can help you understand why a certain file changed

Smart! The confusing part for me is, why is CircleCI not respecting the updated constraints.txt?

**Comment by mcgibbon** [comment]
Found a bug where sklearn training fails if data inputs are mixed types. Will fix.

**Review by mcgibbon — COMMENTED** [review]
Forgot to post these comments yesterday, sorry!

File: external/fv3fit/fv3fit/emulation/data/config.py
Comment (mcgibbon) [reply external/fv3fit/fv3fit/emulation/data/config.py]: Updated.

File: external/fv3fit/fv3fit/keras/_models/convolutional.py
Comment (mcgibbon) [reply external/fv3fit/fv3fit/keras/_models/convolutional.py:97]: Could be, I should explore this more with tensorboard profiling later. I'll leave this umerged here to expedite the process but I'll merge this in a follow-on PR if it passes tests.

Comment (mcgibbon) [reply external/fv3fit/fv3fit/keras/_models/convolutional.py:97]: Ah I just realized this is in the convolutional code, not the dense code. So it wouldn't be the cause of the slowdowns we're currently investigating.

**Comment by mcgibbon** [comment]
I have a local fix for the failing unit test, will push it tomorrow to avoid resetting an image build.

**Pull request merged**
